src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs

```diff
@@ -454,9 +444,43 @@ internal bool IsReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAn
             return AnyConstraintTypes(constraintTypes, inProgress, (type, arg) => ConstraintImpliesReferenceType(type.TypeSymbol, arg));
         }
 
-        internal bool IsNotNullableIfReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAnnotations> constraintTypes, ConsList<TypeParameterSymbol> inProgress)
+        internal bool? IsNotNullableIfReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAnnotations> constraintTypes, ConsList<TypeParameterSymbol> inProgress)
         {
-            return AnyConstraintTypes(constraintTypes, inProgress, (type, arg) => ConstraintImpliesIsNotNullableIfReferenceType(type, arg));
+            Debug.Assert(!constraintTypes.IsDefaultOrEmpty);
+            inProgress = inProgress.Prepend(this);
+
+            bool? result = false;
+            foreach (TypeSymbolWithAnnotations constraintType in constraintTypes)
+            {
+                if (constraintType.IsAnnotated)
+                {
+                    continue;
+                }
+
+                if (constraintType.TypeKind == TypeKind.TypeParameter)
+                {
+                    bool? isNotNullableIfReferenceType = ((TypeParameterSymbol)constraintType.TypeSymbol).GetIsNotNullableIfReferenceType(inProgress);
+
+                    if (isNotNullableIfReferenceType == false)
+                    {
+                        continue;
+                    }
+                    else if (isNotNullableIfReferenceType == null)
+                    {
+                        result = null;
+                        continue;
+                    }
+                }
+
+                if (constraintType.NonNullTypesContext.NonNullTypes == true)
```

Comment:
>NonNullTypesContext [](start = 35, length = 19)

I found this surprising. The NonNullTypes context of the constraint type doesn't seem relevant. I would have thought that the context of the member declaring those constraints would come into play instead.
What would be an example for this?

Reply:
> I found this surprising. The NonNullTypes context of the constraint type doesn't seem relevant.

If context of the constraint type is not relevant, then why do we have it in TypeSymbolWithAnnotations?

>  I would have thought that the context of the member declaring those constraints would come into play instead.

Isn't that what is getting captured in TypeSymbolWithAnnotations.NonNullTypesContext? We are interested in context that was in play when the type reference was bound, whether at constraint declaration site, or at generic construction site.

---
In reply to: [213848658](https://github.com/dotnet/roslyn/pull/29586#discussion_r213848658) [](ancestors = 213848658)